### PR TITLE
libwebsockets: more build options

### PR DESCRIPTION
**Pull request by dosvald**
This updates the libwebsockets package Makefile to expose more options like libuv, CGI and more. To support more options, build variants were abandoned, and config options were made instead.

### CODE REVIEW

**Comment by karlp** [comment]
I hope to look at this this week.  One question, what would the downloadable packages be from the image builders now?  Previously, both the https://downloads.openwrt.org/snapshots/trunk/ar71xx/generic/packages/packages/libwebsockets-cyassl_1.7.5-1_ar71xx.ipk and https://downloads.openwrt.org/snapshots/trunk/ar71xx/generic/packages/packages/libwebsockets-openssl_1.7.5-1_ar71xx.ipk would be available for download and install.  Unless I've misunderstood, with the changes proposed, there'll only be a single download with no SSL?  That was a strong advantage of the build variants.

**Comment by hnyman** [comment]
Buildbot needs variants to be able to build the no-ssl and both ssl versions. With just config options, only the default (no-ssl) is built, like @karlp said.

**Comment by dosvald** [comment]
> with the changes proposed, there'll only be a single download with no SSL? That was a strong advantage of the build variants.

Yes, without variants, there can only be one .ipk available. We could make SSL on by default however, but then no .ipk without SSL support/dependency would be available.

If multiple .ipks are to be supported, I see no other way of supporting more configurable options than adding a new -custom variant with these suboptions, kind of like dnsmasq package in base does it with dnsmasq-full.

**Comment by karlp** [comment]
So, feedback?  Might be worth taking this to lede-dev and openwrt-devel to see what other users of libwebsockets might want?  Bear in mind that currently, you only have SSL enabled, there's no "nossl" option anyway at the moment.  I want SSL to work for the default install, but by all means have all sorts of options if you are building for yourself.

**Comment by dosvald** [comment]
Okay, so should I just change default to have OpenSSL as SSL provider in this line?

https://github.com/openwrt/packages/pull/2784/files#diff-79b2d20017acd94f52ba20cda92b7014R57

**Comment by karlp** [comment]
Or it has to stay as variants.  libustream has variants for all the ssl libraries it supports, it doesn't seem unreasonable to me that ssl is fixed in variants, and most other features are enabled by default, and then everything else is just custom menuconfig building?

**Comment by hnyman** [comment]
+1 for keeping the three SSL variants and having everything else as config options.

**Comment by dosvald** [comment]
So it would have these variants:

- libwebsockets-openssl (as before)
- libwebsockets-cyanssl (as before)
- libwebsockets-nossl
- libwebsockets-custom:
  * which ssl to use (OpenSSL or CyaSSL or no ssl, or hopefully soon-ish mbedTLS/PolarSSL)
  * libuv event loop support yes/no
  * ...

I can make that version...

**Comment by karlp** [comment]
fwiw, LWS 2.x won't compile with wolfssl/cyassl anyway.... and we're going to want to update to lws2.x "soon" anyway.  Something related to options being behind --enable-stunnel, but even with that rebuilt locally, it still fails.  The staging_dir/usr/include/wolfssl/options.h has the HAVE_STUNNEL set, but no part of wolfssl ever includes that, so it needs both a) lede/openwrt changs to the base package to enable STUNNEL _and_ patching lws to include options.h....

**Comment by dosvald** [comment]
Well, I've tried to accommodate SSL variants while adding an option or two, and can't seem to come up with a good solution. I suspected as much, feeling that using both variants and config options will have problems, so that's why my first proposal was to scrap variants altogether.

Anyway, I've tried making the -custom variant with options, as I suggested above.

First, It ended up duplicating stuff in the Makefile, because it needs to enable OpenSSL either if variant tells it to, or if variant is custom and OpenSSL is selected in menu. But Make has no "or" operator in comparisons, so it bloats up the Makefile.

Second, it has problems with the -custom variant satisfying the dependency. We can't have conditional PROVIDES directive like we can for DEPENDS. If we could do that, then -custom would conditionally provide the libwebsockets-openssl if the config symbol selects OpenSSL. Without that, the -custom can only be used if some package explicitly DEPENDS on "libwebsockets-openssl || libwebsockets-custom && CONFIG_libwebsockets-custom_openssl: (if such a construct is even valid). Additionally, the PROVIDES directive seems to make dependent packages hidden as menuconfig doesn't know which one it should select to resolve dependency.

Last but not least, since libwebsockets is a library which has to put headers in the global staging dir, selecting multiple variants results in the staging dir libwebsockets.h being overwritten with whatever selected variant comes up last in menuconfig. I don't know if this makes something break but it sound like it might at some point.

So, I'll try to think of another way...
Meanwhile, can somebody explain to me the benefit of variants, and what use cases they enable?

**Comment by karlp** [comment]
If it helps, given no-one's ever commented on the cyassl version, and given that lws2 doesn't compile with wolfssl in it's current state, and lws doesn't seem inclined to fix it (and neither do I)  I'm tending to simply drop wolfssl.  In theory, LWS supports mbedtls, but it's not tested at this point, so if it simplifies what you're trying to do, then consider LWS as having only one option, and that's OpenSSL.

As for what benefit variants provide, it means that you can do "opkg install package-variant"  Which is an important use case.  Config only options can only be used by people building their own images, not even imagebuilder can use them.

**Comment by dosvald** [comment]
I've updated the PR.

The -full variant hard depends on libopenssl since I couldn't find a good way to have all of that configurable as well as having the -full variant satisfy -openssl version. It now has a fixed DEPENDS on libopenssl and  PROVIDES libwebsockets-openssl. This however makes menuconfig hide mosquitto until I choose whiether to resolve -openssl via -full or via -openssl itself.

Any hints / new feedback?

File: libs/libwebsockets/Makefile [hunk 1]
```diff
@@ -46,29 +36,59 @@ define Package/$(PKG_NAME)/Default
...
 	$(call Package/$(PKG_NAME)/Default)
 	TITLE += (CyaSSL)
 	DEPENDS += +libcyassl
 	VARIANT:=cyassl
 endef
 
+define Package/$(PKG_NAME)-full
+	$(call Package/$(PKG_NAME)/Default)
+	TITLE += (OpenSSL + extras)
+	DEPENDS += +libopenssl \
```
Comment (karlp) [inline libs/libwebsockets/Makefile]: You can use DEPENDS +=, you don't need to hack around with \ and is it even too long for a single line anyway?

**Comment by karlp** [comment]
My comments at this point are:
a) is there anyone actually looking for a -nossl variant at all?
b) can the libuv be inverted?  Ie, instead of a config option to enable it, have it be enabled, _if_ the user has selected libuv?  
c) why not just include cgi all the time?  how much difference does it make?  (And is it even meaningful for the 1.7.x branch?)  even if you go down busybox path and customize _everything_ it's only meaningful to have a relatively complete default install.

File: libs/libwebsockets/Makefile (hunk 1 above)
Comment (dosvald) [inline libs/libwebsockets/Makefile]: Done

**Comment by dosvald** [comment]
Regarding a) not me, but maybe you could have mosquitto nossl have ws feature that way

b) Right now it's default-on with conditional depends. Did I understand right, you're suggesting no depends, and auto detect with no config option? What about opkg installing the package then, It won't bring in the libuv dependency even if it was built with libuv?

c) I put it configurable because It's default-off in the library itself. I'll look into the numbers for size.

**Comment by dosvald** [comment]
Additonal (size) info for c)
```
 52K cgi/libwebsockets-full_2.0.2-2_ar71xx.ipk
120K cgi/libwebsockets.so.8
 48K nocgi/libwebsockets-full_2.0.2-2_ar71xx.ipk
112K nocgi/libwebsockets.so.8
```

**Comment by dosvald** [comment]
Any news?
Could we bring in libuv feature + dependency, and leave SSL options as-is?

**Comment by karlp** [comment]
Now that we're up to 2.x, perhaps a new PR with less of this clutter around SSL might be nice to follow for that.  Perhaps as a package adding lwsws as a separate package maybe?

**Comment by dosvald** [comment]
Did you mean something like this:

New variant libwebsockets-full = libwebsockets-openssl + all options for
lwsws.
Then package lwsws depends on -full and that's it.

What do you think?

**Comment by karlp** [comment]
sounds reasonable at first glace, yeah :)

**Pull request closed without merging**